frontend/javascripts/oxalis/view/right-border-tabs/metadata_table.tsx

```diff
@@ -0,0 +1,167 @@
+  item,
+  setMetadata,
+  readOnly,
+}: {
+  item: ItemType;
+  setMetadata: (item: ItemType, newProperties: MetadataEntry[]) => void;
+  readOnly: boolean;
+}) {
+  const updateMetadataEntryByIndex = (
+    item: ItemType,
+    index: number,
+    newPropPartial: Partial<MetadataEntry>,
+  ) => {
+    if (readOnly) {
+      return;
+    }
+    const newProps = item.metadata.map((element, idx) =>
+      idx === index
+        ? {
+            ...element,
+            ...newPropPartial,
+          }
+        : element,
+    );
+
+    setMetadata(item, newProps);
+  };
+
+  const removeMetadataEntryByIndex = (item: ItemType, index: number) => {
+    const newProps = item.metadata.filter((_element, idx) => idx !== index);
+    setMetadata(item, newProps);
+  };
+
+  const addMetadataEntry = (item: ItemType, newProp: MetadataEntry) => {
+    const newProps = item.metadata.concat([newProp]);
+    setMetadata(item, newProps);
+  };
+
+  const getDeleteEntryButton = (_: APIMetadataWithError, index: number) => (
+    <div className="flex-center-child">
+      <Button
+        type="text"
+        disabled={readOnly}
+        style={{ width: 16, height: 19 }}
+        icon={
+          <CloseOutlined
+            style={{
+              color: "var(--ant-color-text-tertiary)",
+              width: 16,
+            }}
+          />
+        }
+        onClick={() => {
+          removeMetadataEntryByIndex(item, index);
+        }}
+      />
+    </div>
+  );
+
+  const getKeyInput = (record: APIMetadataWithError, index: number) => {
+    return (
+      <div style={{ display: "flex" }}>
+        <InputWithUpdateOnBlur
+          prefix={<TagsOutlined />}
+          className="transparent-input"
+          value={record.key}
+          disabled={readOnly}
+          onChange={(value) => updateMetadataEntryByIndex(item, index, { key: value })}
+          placeholder="Property"
+          size="small"
+          validate={(value: string) => {
+            if (
+              // If all items (except for the current one) have another key,
+              // everything is fine.
+              item.metadata.every((otherItem, idx) => idx === index || otherItem.key !== value)
+            ) {
+              return null;
+            }
+            return "Each key must be unique and can only be used once.";
+          }}
+          id={getKeyInputIdForIndex(index)}
+        />
+      </div>
+    );
+  };
+
+  const itemMetadata = item.metadata.map((prop) => ({
+    key: prop.key,
+    type:
+      prop.stringValue != null
+        ? APIMetadataEnum.STRING
+        : prop.numberValue != null
+          ? APIMetadataEnum.NUMBER
+          : APIMetadataEnum.STRING_ARRAY,
+    value: prop.stringValue || prop.numberValue || prop.stringListValue || "",
+  }));
+
+  const getValueInput = (record: APIMetadataWithError, index: number) => {
+    return (
+      <MetadataValueInput
+        record={record}
```

Comment:
I renamed it to `Entry`. However, the bigger problem is that we have the following two types:

```
export type APIMetadata = {
  type: APIMetadataEnum;
  key: string;
  value: string | number | string[];
};

export type MetadataEntry = {
  key: string;
  stringValue?: string;
  boolValue?: boolean;
  numberValue?: number;
  stringListValue?: string[];
};
```

since their structure differs (because of protobuf vs sql), we have two types. how should we name them? I have trouble coming up with a good name, because one would need to find a super set for dataset+folder and one for segments+trees...

cc @fm3

Reply:
In the backend I started using `MetadataEntryProto` for the latter. I’m not a huge fan of this solution but I don’t have anything better at this time